rpc/client.ts

```diff
@@ -73,57 +70,82 @@ export class Client<
-    this.pendingCalls[message.id] = waiter
-    this.providerRef.send(message)
+    this.pendingCalls[id] = waiter
+    this.providerRef.send({
+      jsonrpc: "2.0",
+      id,
+      method,
+      params,
+    })
     return waiter
   }
 
-  subscribe: ClientSubscribe<SendErrorData, HandlerErrorData> = (message, listener) => {
-    const waiter = deferred<string | undefined>()
-    const stop = () => {
-      delete this.pendingSubscriptions[message.id]
-      const activeSubscriptionId = this.activeSubscriptionByMessageId[message.id]
-      if (activeSubscriptionId) {
-        delete this.activeSubscriptions[activeSubscriptionId]
+  subscriptionFactory = <
+    Params extends unknown[] = any[],
+    NotificationData = any,
+  >(): SubscriptionFactory<Params, NotificationData, SendErrorData, HandlerErrorData> =>
+    (
+      subscribeMethod,
+      unsubscribeMethod,
+      params,
+      createListener,
+    ) => {
+      const id = this.providerRef.nextId()
+      const waiter = deferred<any>() // TODO: type this narrowly
+      const stop = async (value?: unknown) => {
+        delete this.pendingSubscriptions[id]
+        const activeSubscriptionId = this.activeSubscriptionByMessageId[id]
+        if (activeSubscriptionId) {
+          delete this.activeSubscriptions[activeSubscriptionId]
+          await this.call(
+            this.providerRef.nextId(),
+            unsubscribeMethod,
+            [activeSubscriptionId],
+          )
+        }
+        delete this.activeSubscriptionByMessageId[id]
+        waiter.resolve(value)
       }
-      delete this.activeSubscriptionByMessageId[message.id]
-      waiter.resolve(activeSubscriptionId)
-    }
-    const listenerBound = listener.bind({
-      message,
-      stop,
-      state: <T>(ctor: new() => T) => {
-        return getOrInit(
-          getOrInit(this.subscriptionStates, message.id, () => new WeakMap()),
-          ctor,
-          () => new ctor(),
-        )
-      },
-    }) as ClientSubscribeListener<SendErrorData, HandlerErrorData>
-    this.pendingSubscriptions[message.id] = (maybeError) => {
-      listenerBound(maybeError)
-      if (maybeError instanceof Error) {
-        stop()
+      const listener = createListener({
+        end(value = undefined!) {
+          return new U.End(value)
+        },
+        endIfError(value) {
+          return (value instanceof Error ? new U.End(value) : undefined) as any
+        },
+      })
+      const maybeStop = (maybeEnd: unknown) => {
+        if (maybeEnd instanceof U.End) {
+          stop(maybeEnd.value)
+        } else if (maybeEnd instanceof Error) {
+          stop(maybeEnd)
+        }
       }
-    }
-    this.call(message)
-      .then((maybeError) => {
+      this.pendingSubscriptions[id] = (event) => {
+        const result = listener(event)
+        if (result instanceof Promise) {
+          result.then(maybeStop)
+        } else {
+          maybeStop(result)
+        }
+      }
```

Comment:
If `event` is an `Error` and `listener` doesn't returns it, the subscription won't stop.

Perhaps `maybeStop` could be updated to 


```
      const maybeStop = (maybeEnd: unknown, event: unknown) => {
        if (maybeEnd instanceof U.End) {
          stop(maybeEnd.value)
        } else if (maybeEnd instanceof Error) {
          stop(maybeEnd)
        } else if (event instanceof Error) {
          stop(event)
        }
      }
```